### DISCUSSION TITLE
How to display pdf offline in Xamarin.Forms apps?

### DISCUSSION THREAD

**Discussion by GoldenGlitterBug** [xamarin/Xamarin.Forms#15099] [Q&A]
Using [PDFJS](https://mozilla.github.io/pdf.js/getting_started/#download) for displaying PDF files in a Xamarin.Forms apps. Copied the PDFJS folder and PDF files to the android assets folder. I am able to view/display PDFs on Android and iOS when online (connected to internet or mobile service). 

But when I am offline I get following error:
`"Access to XMLHttpRequest at 'file:///android_asset/abcd.pdf' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, chrome, https.", source: file:///android_asset/pdfjs/web/viewer.js`

My code:

page.xaml
```
<WebView
    x:Name="webView"
    BackgroundColor="White"
    HorizontalOptions="FillAndExpand"
    VerticalOptions="FillAndExpand"
    Margin="20">           
</WebView>
```

page.xaml.cs
```
if (Device.RuntimePlatform == Device.Android)
{
      string localPath = "file:///android_asset/abcd.pdf";
      webView.Source = $"file:///android_asset/pdfjs/web/viewer.html?file={WebUtility.UrlEncode(localPath)}";
}
else
     webView.Source = Path.Combine(DependencyService.Get<IBaseUrl>().Get(), "abcd.pdf"); 
```

PdfWebViewRenderer.cs
```
[assembly: ExportRenderer(typeof(PdfWebView), typeof(PdfWebViewRenderer))]
namespace MobileApp.Droid.Scripts
{
	public class PdfWebViewRenderer : WebViewRenderer
	{
		public PdfWebViewRenderer(Context context) : base(context)
		{
		}

		protected override void OnElementChanged(ElementChangedEventArgs<WebView> e)
		{
			base.OnElementChanged(e);

			if (e.NewElement != null)
			{
				Control.Settings.AllowFileAccess = true;
				Control.Settings.AllowFileAccessFromFileURLs = true;
				Control.Settings.AllowUniversalAccessFromFileURLs = true;
			}
		}

		// If you want to enable scrolling in WebView uncomment the following lines.
		//public override bool DispatchTouchEvent(MotionEvent e)
		//{
		//    Parent.RequestDisallowInterceptTouchEvent(true);
		//    return base.DispatchTouchEvent(e);
		//}
	}
}
```

I am using pdf.js (https://github.com/timvandermeij/pdf.js) October 2020 version. Does this work Offline when mobile not connected to internet? My requirement is to display PDF from asset folder 'file:///android_asset/abcd.pdf' when user is Offline..

Also when I try to download the latest stable [version(v2.12.313)](https://mozilla.github.io/pdf.js/getting_started/#download) or any other previous version I get lots of errors. 

How to resolve this error or display PDF Offline in Xamarin.Forms Android and iOS apps?

**Comment by jingliancui** [comment]
For Android you can use this https://www.nuget.org/packages/XamarinLibrary.Xamarin.Android.Github.Barteksc.AndroidPdfViewer/

For iOS you can use pdfkit.

**Comment by GoldenGlitterBug** [comment]
Thanks. From online reviews it's not very efficient.

Namarata Chaturvedi

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
